compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -118,22 +118,20 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
     }
 
     pub fn deref<Cx: LayoutTypeMethods<'tcx>>(self, cx: &Cx) -> PlaceRef<'tcx, V> {
+        if self.layout.ty.is_box() {
+            bug!("dereferencing {:?} in codegen", self.layout.ty);
+        }
+
         let projected_ty = self
             .layout
             .ty
             .builtin_deref(true)
```

Comment:
Maybe `builtin_deref` should take a mode `enum` like this:
```rust
enum BuiltinDerefMode {
    /// `Box<T>`, `&T`, and optionally `*T`.
    UserFacing {
        /// Whether to also allow `*T`.
        allow_unsafe: bool
    },

    /// `&T` and `*T`.
    LowLevel,
}
```

(Though `LowLevel` should probably be named something better)